pydantic/_internal/_generate_schema.py

```diff
@@ -768,23 +769,68 @@ def apply_validators(schema: core_schema.CoreSchema, validators: list[Validator]
     return schema
 
 
+_SerializerType = Literal[
+    'field-wrap',
+    'general-wrap',
+    'field-plain',
+    'general-plain',
+]
+
+
+def _is_unbound_instance_method(sig: Signature, func: Callable[..., Any]) -> bool:
+    if len(sig.parameters) < 2:
+        raise TypeError(f'Unrecognized serializer signature for {func.__name__}: {sig}')
+    first = next(iter(sig.parameters.keys()))
+    return sig.parameters[first].name == 'self'
+
+
+def _infer_serializer_type_from_signature(sig: Signature, func: Callable[..., Any]) -> _SerializerType:
+    if len(sig.parameters) == 2:
+        # (value, info) -> Any
+        return 'general-plain'
+    elif len(sig.parameters) == 3:
+        # (self, value, info) -> Any or (value, nxt, info)
+        if _is_unbound_instance_method(sig, func):
+            return 'field-plain'
+        return 'general-wrap'
+    elif len(sig.parameters) == 4 and _is_unbound_instance_method(sig, func):
+        # (self, value, nxt, info) -> Any
+        return 'field-wrap'
+    raise TypeError(f'Unrecognized serializer signature for {func.__name__}: {sig}')
```

Comment:
I think we should include information in this error for users about how they can find what the "recognized" serializer signatures are. Something like: `See pydantic.SerializerCallableProtocol` (or whatever it is called)

Reply:
I added text describing the signatures. We don't have a single type, it's a union, and it's exported from `pydantic-core`. The whole callable protocol thing is also a bit confusing. So I decided to add a text explanation rather than pointing at a type. In the future when we have this in our docs we should just link to that.

Reply:
> I added text describing the signatures.

Has this been pushed yet?

I think the important thing for me as a user would be that if this TypeError gets raised, I immediately know what I need to do to resolve it (don't need to start opening pydantic source code or Google). I don't know what the best description is, understand if it's not referencing the Protocol stuff, but what I'm getting at is that a user should be able to look at just a stacktrace and know how to proceed from there.

Reply:
https://github.com/pydantic/pydantic/blob/6cf49f95f453e24241f4fda744b51af523b1a177/pydantic/_internal/_generate_schema.py#L780-L806

Reply:
Sorry, I think github was confusing me by showing me older state when I clicked on some item above, and I didn't realize I wasn't seeing the latest. Anyway this definitely looks good.